### PR TITLE
[main] Update fix for trailing stash and add tests

### PR DESCRIPTION
**Pull request by suksw**
### Purpose
- A minor fix for a bug introduced when adding the change https://github.com/wso2/product-microgateway/pull/2837
- Add tests 
  - Invoking path with a trailing slash not given in the initial API definition
  - Invoke rewritten path (via API policy) with a trailing slash
  - Invoke default and non-default APIs with a trailing slash
  - Invoke `/context/*` for all methods when path is /*
  - Invoke `/context` for a few methods (get and post) when path is /*

### Issues
Related to https://github.com/wso2/product-microgateway/issues/2864

### Automation tests
 - Unit tests added: Updated
 - Integration tests added: Yes

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: integration/test-integration/src/test/resources/openAPIs/trailing_slash_openAPI.yaml [hunk 1]
```diff
@@ -0,0 +1,73 @@
+openapi: "3.0.0"
+info:
+  version: 1.0.0
+  title: Trailing Slash
+servers:
+  - url: http://mockBackend:2383/v2
+x-wso2-basePath: /trailing-slash
+paths:
+  /echo-full/findByStatus:
```
Comment (VirajSalaka) [inline integration/test-integration/src/test/resources/openAPIs/trailing_slash_openAPI.yaml:9]: Shall we check with another resource with a trailing slash and see the behavior?
ex: `/foo/ `

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go [hunk 2]
```diff
@@ -1391,5 +1391,5 @@ func getDefaultResourceMethods(apiType string) []string {
 
 func getDefaultVersionBasepath(basePath string, version string) string {
 	context := strings.ReplaceAll(basePath, "/"+version, "")
```
Comment (VirajSalaka) [inline adapter/internal/oasparser/envoyconf/routes_with_clusters.go:1393]: Shall we fix this logic such that only the trailing part will be swapped?
I think we could use LastIndexOf and then pick a substring.

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go [hunk 3]
```diff
@@ -1391,5 +1391,5 @@ func getDefaultResourceMethods(apiType string) []string {
...
-	return fmt.Sprintf("(%s|%s)", basePath, context)
+	return fmt.Sprintf("(?:%s|%s)", basePath, context)
```
Comment (VirajSalaka) [inline adapter/internal/oasparser/envoyconf/routes_with_clusters.go:1394]: Can you please explain the purpose behind `?`

Comment (suksw) [reply adapter/internal/oasparser/envoyconf/routes_with_clusters.go:1394]: This is to make "(%s|%s)" behave as a non-capturing group in the regex. Behavior we saw where the basepath got placed again in the locations of the path parameters, gets resolved when we add this. I'll add a comment here.

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go (hunk 2 above)
Comment (suksw) [reply adapter/internal/oasparser/envoyconf/routes_with_clusters.go:1393]: This is the `getDefaultVersionBasepath`. This is correct right?

File: integration/test-integration/src/test/resources/openAPIs/trailing_slash_openAPI.yaml (hunk 1 above)
Comment (suksw) [reply integration/test-integration/src/test/resources/openAPIs/trailing_slash_openAPI.yaml:9]: I updated policy and default API testcases as you suggested. They call other paths that are not `/echo-full`. Changes made to Default API testcase covers all three cases. https://github.com/wso2/product-microgateway/pull/2878/commits/fc9d1952fb8ad3d82c7fd091e0ff550193152fa6 Will these do?

File: integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/apiDeploy/AllHttpMethodsForWildcardTestCase.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+    @BeforeClass(description = "Get Prod token")
+    void start() throws Exception {
+        jwtTokenProd = TokenUtil.getJwtForPetstore(TestConstant.KEY_TYPE_PRODUCTION, null, false);
+        headers.put(HttpHeaderNames.AUTHORIZATION.toString(), "Bearer " + jwtTokenProd);
+    }
+
+    @Test(description = "test path with trailing slash for GET")
+    public void testTrailingSlashGET() throws Exception {
+        EchoResponse echoResponse = Utils.invokeEchoGet(API_CONTEXT, "/",
```
Comment (pubudu538) [inline integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/apiDeploy/AllHttpMethodsForWildcardTestCase.java:50]: I guess this test is for invoking /api_context/. We need another test case for /api_context. There is no end slash.

Comment (pubudu538) [reply integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/apiDeploy/AllHttpMethodsForWildcardTestCase.java:50]: In APIM if you create an API with /* resource then you can invoke this as /apicontext/ and /apicontext. Can't we achieve the same here?

Comment (VirajSalaka) [reply integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/apiDeploy/AllHttpMethodsForWildcardTestCase.java:50]: we should fix it then

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go (hunk 3 above)
Comment (VirajSalaka) [reply adapter/internal/oasparser/envoyconf/routes_with_clusters.go:1394]: Great. Thanks for the explaination.

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go (hunk 2 above)
Comment (VirajSalaka) [reply adapter/internal/oasparser/envoyconf/routes_with_clusters.go:1393]: Let's assume the basePath is `/foo/v2` and version is `/v2`. So the complete path would be `/foo/v2/v2` in that case, v2 would be replaced twice right?
This is an edge case actually, but I have seen people doing that still.

File: integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/apiDeploy/AllHttpMethodsForWildcardTestCase.java (hunk 4 above)
Comment (suksw) [reply integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/apiDeploy/AllHttpMethodsForWildcardTestCase.java:50]: This works. Let me update the tests. Didn't work while writing tests. Must have written the test wrong.

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go (hunk 2 above)
Comment (suksw) [reply adapter/internal/oasparser/envoyconf/routes_with_clusters.go:1393]: +1

File: integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/apiDeploy/AllHttpMethodsForWildcardTestCase.java (hunk 4 above)
Comment (suksw) [reply integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/apiDeploy/AllHttpMethodsForWildcardTestCase.java:50]: fixed with https://github.com/wso2/product-microgateway/pull/2878/commits/10bdb144f96c4a1ffcea84c43ab7803b61415a9a

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go (hunk 2 above)
Comment (suksw) [reply adapter/internal/oasparser/envoyconf/routes_with_clusters.go]: fixed with https://github.com/wso2/product-microgateway/pull/2878/commits/3fb46d23794f856dd8102d2b17b8172ab0f13a37

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go (hunk 3 above)
Comment (suksw) [reply adapter/internal/oasparser/envoyconf/routes_with_clusters.go:1399]: Fixed with https://github.com/wso2/product-microgateway/pull/2878/commits/c9b1c53fe6698d6413e3057d3915dd31808bf1df

File: integration/test-integration/src/test/resources/openAPIs/trailing_slash_openAPI.yaml (hunk 1 above)
Comment (VirajSalaka) [reply integration/test-integration/src/test/resources/openAPIs/trailing_slash_openAPI.yaml:9]: The idea here is to like verify if the swagger resource has a trailing `/` and when we are forwarding it to the backend, whether we append that remaining slash.
As I saw, all the tests are like there is a resource with no slash and we are checking the both options. But nothing covers the pattern I tell right?

Comment (suksw) [reply integration/test-integration/src/test/resources/openAPIs/trailing_slash_openAPI.yaml:9]: Thanks for the explanation. Got the idea. Will update.

Comment (suksw) [reply integration/test-integration/src/test/resources/openAPIs/trailing_slash_openAPI.yaml]: fixed with https://github.com/wso2/product-microgateway/pull/2878/commits/dc1a2a713cc236684b6626c495fd64319604069a

**Review by VirajSalaka — APPROVED** [review]
